### PR TITLE
Update CAT create/edit view to support creation and modification of credit courses

### PR DESCRIPTION
**Pull request by rlucioni**
XCOM-511

@clintonb and @AlasdairSwan, this is still a WIP, but I'd appreciate an early look, if you have the time. I'm still trying to get the table validation to work properly, and I still need to write tests.

### CODE REVIEW

File: ecommerce/static/js/views/course_form_view.js [hunk 1]
```diff
@@ -263,19 +261,18 @@ define([
...
                             viewClass,
                             view = this.courseSeatViews[seatType];
 
                         if (!view) {
-                            model = this.model.getOrCreateSeat(seatType);
+                            seats = new ProductCollection(this.model.getOrCreateSeats(seatType));
                             viewClass = this.courseSeatViewMappings[seatType];
 
-                            if (viewClass && model) {
+                            if (viewClass && seats.length >= 1) {
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_form_view.js]: Nitpick: > 0 would mean you can reduce the code by 1 character

File: ecommerce/static/js/views/course_form_view.js [hunk 2]
```diff
@@ -311,6 +308,8 @@ define([
                 this.$el.find('.alerts').append(view.el);
                 this.alertViews.push(view);
 
+                $('body').animate({ scrollTop: 0 }, 'fast');
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_form_view.js]: This should really be handled at the page level because a view should only interact with its own elements. Is scrolling to the top of the page a common occurrence?

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js [hunk 3]
```diff
@@ -35,55 +41,110 @@ define([
             },
 
             className: function () {
-                return 'row course-seat ' + this.seatType;
+                var className = 'row ' + this.seatType;
+
+                if (this.seatType != 'credit') {
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: Please use !==

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js [hunk 4]
```diff
@@ -35,55 +41,110 @@ define([
...
+                // Instantiate new Views handling data binding for each Model in the Collection.
+                this.collection.each( function (seat) {
+                    row = new self.rowView({
+                        model: seat,
+                        isRemovable: false,
+                        parent: self.$el.find('tbody'),
+                        course: self.course,
+                        seatType: self.seatType
+                    });
+                });
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: You can pass context in here so you don't need to alias this to the self var.

```
this.collection.each( function(seat) {
    ...
}, this);
```

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js [hunk 5]
```diff
@@ -35,55 +41,110 @@ define([
...
                 return this;
             },
 
-            cleanIdVerificationRequired: function (val) {
-                return _s.toBoolean(val);
-            },
+            /**
+             * Renders a table of course seats sharing a common seat type.
+             */
+            renderSeatTable: function () {
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: There should not be a space between function and (
It should be

```
renderSeatTable: function( {
```

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js [hunk 6]
```diff
@@ -35,55 +41,110 @@ define([
...
+
+                // Inject the wrapper template.
+                this.$el.html(this.template());
+
+                // Instantiate new Views handling data binding for each Model in the Collection.
+                this.collection.each( function (seat) {
+                    row = new self.rowView({
+                        model: seat,
+                        isRemovable: false,
+                        parent: self.$el.find('tbody'),
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: You are using this selector multiple times so it makes sense to assign it to a variable before you go into the each loop

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js [hunk 7]
```diff
@@ -35,55 +41,110 @@ define([
...
-             * Return the input data from the form fields.
+            /**
+             * Adds a new row to the seat table.
+             *
+             * @param e
              */
-            getData: function () {
-                var data = {},
-                    fields = ['certificate_type', 'id_verification_required', 'price', 'expires'];
+            addSeatTableRow: function (e) {
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: e isn't being used, please remove

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+                // Add seat model to course product collection. This operation
+                // is idempotent.
+                this.course.get('products').add(this.model);
+
+                this.render();
+            },
+
+            render: function () {
+                this.$el.html(this.template(_.extend({}, this.model.attributes, {isRemovable: this.isRemovable})));
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js]: Please break this line down, there's too much going on for a single line.

File: ecommerce/static/templates/credit_course_seat_form_field_row.html [hunk 9]
```diff
@@ -0,0 +1,28 @@
+<td class="credit-provider">
+    <div class="input-group">
+        <div class="input-group-addon"><i class="fa fa-building-o" aria-hidden="true"></i></div>
+        <input type="text" id="credit_provider" name="credit_provider" class="form-control" value="">
```
Comment (AlasdairSwan) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: All inputs should have associated labels (for a11y)

File: ecommerce/static/templates/credit_course_seat_form_field_row.html [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    </div>
+</td>
+<td class="expires">
+    <div class="input-group">
+        <div class="input-group-addon"><i class="fa fa-calendar" aria-hidden="true"></i></div>
+        <input type="datetime-local" id="expires" name="expires" class="form-control" value="">
+    </div>
+</td>
+<% if (isRemovable) { %>
+<td><span class="table-remove fa fa-trash-o"></span></td>
```
Comment (AlasdairSwan) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: Is this a CTA? If so it should be a button. Need to add screen reader only text explaining the action

File: ecommerce/static/templates/credit_course_seat_form_field_row.html [hunk 11]
```diff
@@ -0,0 +1,28 @@
+<td class="credit-provider">
+    <div class="input-group">
+        <div class="input-group-addon"><i class="fa fa-building-o" aria-hidden="true"></i></div>
```
Comment (AlasdairSwan) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: Is this a CTA? If so it should be a button. Need to add screen reader only text explaining the action. This comment is true of all subsequent input-group-addon instances in the rest of the file.

File: ecommerce/static/js/models/course_seats/credit_seat.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+                        required: true,
+                        msg: gettext('All credit seats must have a credit provider.')
+                    },
+                    credit_hours: {
+                        required: true,
+                        msg: gettext('All credit seats must designate a number of credit hours.')
+                    }
+                }
+            )
+        }, {seatType: 'credit'});
```
Comment (AlasdairSwan) [inline ecommerce/static/js/models/course_seats/credit_seat.js:32]: Why is this a separate object?

File: ecommerce/static/js/views/course_form_view.js (hunk 2 above)
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_form_view.js]: You should also move focus to the top of the page for a11y

File: ecommerce/static/templates/credit_course_seat_form_field.html [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+                    <th><%= gettext('Price (USD)') %></th>
+                    <th><%= gettext('Credit Hours') %></th>
+                    <th><%= gettext('Upgrade Deadline') %></th>
+                    <th><span class="table-add fa fa-plus"></span></th>
+                </tr>
+            </thead>
+            <tbody></tbody>
+        </table>
+        <!-- NOTE: This help-block is here for validation messages. -->
+        <span class="help-block"></span>
```
Comment (AlasdairSwan) [inline ecommerce/static/templates/credit_course_seat_form_field.html]: You should have an aria-live property so screenreaders catch any validation messages (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions)

File: ecommerce/static/js/models/course_seats/course_seat.js [hunk 14]
```diff
@@ -10,6 +10,8 @@ define([
                 expires: null,
                 id_verification_required: null,
                 price: 0,
+                credit_provider: null,
+                credit_hours: 0,
```
Comment (clintonb) [inline ecommerce/static/js/models/course_seats/course_seat.js]: This should default to null, otherwise non-credit seats will be saved with 0 credit hours.

File: ecommerce/static/js/models/course_seats/credit_seat.js [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+    ],
+    function (CourseSeat) {
+        'use strict';
+
+        return CourseSeat.extend({
+            defaults: _.extend({}, CourseSeat.prototype.defaults,
+                {
+                    certificate_type: 'credit',
+                    id_verification_required: true,
+                    price: null,
```
Comment (clintonb) [inline ecommerce/static/js/models/course_seats/credit_seat.js]: Price should default to 0.

File: ecommerce/static/js/models/course_seats/credit_seat.js (hunk 12 above)
Comment (clintonb) [inline ecommerce/static/js/models/course_seats/credit_seat.js:32]: This was a workaround for the fact that I could not create a static/class variable. It didn't work, and should be removed.

Comment (clintonb) [inline ecommerce/static/js/models/course_seats/credit_seat.js:32]: Correction: it does work. Leave it for now, until I can remember why I didn't use it. Weird.

Comment (clintonb) [inline ecommerce/static/js/models/course_seats/credit_seat.js:32]: Some background: https://taurenmills.wordpress.com/2011/10/08/backbone-js-with-class-properties/

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js [hunk 16]
```diff
@@ -19,6 +21,10 @@ define([
             seatType: null,
             template: null,
 
+            events: {
```
Comment (clintonb) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: Any particular reason this wasn't moved to a separate view class?

File: ecommerce/static/js/views/course_form_view.js (hunk 1 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_form_view.js]: Updated.

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js (hunk 3 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: Updated to use `_.isEqual()` so we don't have to think about `!=` as opposed to `!==`.

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+                this.parent.append(this.$el);
```
Comment (clintonb) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js]: The parent should control the children. The parent should call this view's render method and handle the appending. The parent should also call this view's remove method to handle cleanup.

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js (hunk 4 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: Thanks for the tip. Updated.

File: ecommerce/static/templates/credit_course_seat_form_field.html [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+    <div class="form-group">
+        <table class="table table-striped credit-seats">
+            <thead>
+                <tr>
+                    <th><%= gettext('Credit Provider') %></th>
+                    <th><%= gettext('Price (USD)') %></th>
+                    <th><%= gettext('Credit Hours') %></th>
+                    <th><%= gettext('Upgrade Deadline') %></th>
+                    <th><span class="table-add fa fa-plus"></span></th>
```
Comment (clintonb) [inline ecommerce/static/templates/credit_course_seat_form_field.html]: Use `<button>` instead of `<span>` and you won't need to deal with changing the CSS pointer. Also, accessibility.

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js (hunk 5 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: Every function definition in ecommerce currently puts a space between `function` and the opening parentheses. Updating this now will make the diff really dirty. Is it okay if we address this in a separate PR?

Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: Sure

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js (hunk 6 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: Updated.

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js (hunk 7 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: Updated.

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js (hunk 8 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js]: Updated.

File: ecommerce/static/templates/credit_course_seat_form_field_row.html (hunk 9 above)
Comment (rlucioni) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: Updated.

File: ecommerce/static/templates/credit_course_seat_form_field.html (hunk 18 above)
Comment (rlucioni) [inline ecommerce/static/templates/credit_course_seat_form_field.html]: Updated.

File: ecommerce/static/templates/credit_course_seat_form_field_row.html (hunk 10 above)
Comment (rlucioni) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: Updated.

File: ecommerce/static/templates/credit_course_seat_form_field_row.html (hunk 11 above)
Comment (rlucioni) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: This is not a CTA; it's a decorative icon. Each of these input groups is now labelled by using `aria-labelledby` to link input elements to the headers. Is screen reader-only text still required?

Comment (AlasdairSwan) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: Nope, that should be good.

File: ecommerce/static/templates/credit_course_seat_form_field.html (hunk 13 above)
Comment (rlucioni) [inline ecommerce/static/templates/credit_course_seat_form_field.html]: Updated with `aria-live="assertive"`.

Comment (AlasdairSwan) [inline ecommerce/static/templates/credit_course_seat_form_field.html]: It should be polite, because manners are important (but also it will wait until what is being currently read is finished before alerting the user)

File: ecommerce/static/js/models/course_seats/course_seat.js (hunk 14 above)
Comment (rlucioni) [inline ecommerce/static/js/models/course_seats/course_seat.js]: Nice catch. Updated.

File: ecommerce/static/js/models/course_seats/credit_seat.js (hunk 15 above)
Comment (rlucioni) [inline ecommerce/static/js/models/course_seats/credit_seat.js]: Updated.

File: ecommerce/static/templates/credit_course_seat_form_field.html (hunk 13 above)
Comment (rlucioni) [inline ecommerce/static/templates/credit_course_seat_form_field.html]: Cool, I wasn't sure which was preferred for validation messages.

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js (hunk 16 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: Given that this view is responsible for rendering and otherwise controlling tables, it seemed appropriate here.

Comment (clintonb) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: The base CourseSeatFormFieldView doesn't need to know about credit. It would be cleaner to make a CreditCourseSeatFormFieldView (or a more appropriate name).

Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: The base CourseSeatFormFieldView doesn't actually know anything about credit. However, it does know how to manage course seat tables in general. I take advantage of this by extending the base CourseSeatFormFieldView to make a CreditCourseSeatFormFieldView which can render a table of credit seats.

Comment (clintonb) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: The base class has no notion of tables or multiple seats. This should be moved to the credit view class since that's the only place where it is used.

File: ecommerce/static/js/views/course_form_view.js (hunk 2 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_form_view.js]: Updated.

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js (hunk 16 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js]: @clintonb updated. Logic for handling seat tables is now only present in the CreditCourseSeatFormFieldView.

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js (hunk 17 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js]: Updated so that the parent calls its children's render methods and handles appending to the DOM. In light of our earlier conversation, the children still handle their own deletion and cleanup.

**Comment by rlucioni** [comment]
@clintonb and @AlasdairSwan, this is ready for final review.

File: ecommerce/static/js/utils/utils.js [hunk 19]
```diff
@@ -81,6 +85,43 @@ define([
...
+            /**
+             * Bind the provided view for form validation.
+             *
+             * @param {Backbone.View} view
+             */
+            bindValidation: function (view) {
+                /* istanbul ignore next */
+                Backbone.Validation.bind(view, {
+                    valid: function (view, attr) {
+                        var $el = view.$('[name=' + attr + ']'),
```
Comment (AlasdairSwan) [inline ecommerce/static/js/utils/utils.js]: Shouldn't this be

```
view.$el.find('[name=' + attr + ']'),
```

File: ecommerce/static/js/utils/utils.js [hunk 20]
```diff
@@ -81,6 +85,43 @@ define([
...
+                            className = 'invalid-' + attr,
+                            $msg = $helpBlock.find('.' + className);
+
+                        $msg.remove();
+
+                        $group.removeClass('has-error');
+                        $helpBlock.addClass('hidden');
+                    },
+                    invalid: function (view, attr, error) {
+                        var $el = view.$('[name=' + attr + ']'),
```
Comment (AlasdairSwan) [inline ecommerce/static/js/utils/utils.js]: Shouldn't this be

```
view.$el.find('[name=' + attr + ']'),
```

File: ecommerce/static/js/views/course_form_view.js [hunk 21]
```diff
@@ -306,11 +310,19 @@ define([
              * @param {Sring} message - Message to display to the user.
              */
             renderAlert: function (level, message) {
-                var view = new AlertView({level: level, title: gettext('Error!'), message: message});
+                var view = new AlertView({level: level, title: gettext('Error!'), message: message}),
+                    $alerts = this.$el.find('.alerts');
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_form_view.js]: If there are multiple alerts possible without re-rendering the view it could be worthwhile setting a view variable this.$alerts after render to avoid having to create the jQuery object every time an alert has to be rendered

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+                this.course = options.course;
+                this.isRemovable = options.isRemovable;
+
+                this.render();
+            },
+
+            render: function () {
+                var context = _.extend({}, this.model.attributes, {isRemovable: this.isRemovable});
+
+                this.$el.html(this.template(context));
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js:50]: Doesn't passing an extension of this.model instead of the actual model mess-up the two-way data-binding that stickit is doing?

File: ecommerce/static/js/utils/utils.js (hunk 19 above)
Comment (rlucioni) [inline ecommerce/static/js/utils/utils.js]: Updated.

File: ecommerce/static/js/utils/utils.js (hunk 20 above)
Comment (rlucioni) [inline ecommerce/static/js/utils/utils.js]: Updated.

File: ecommerce/static/js/views/course_form_view.js (hunk 21 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_form_view.js]: Updated.

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js (hunk 22 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js:50]: @AlasdairSwan I don't think so. By default, stickit uses `view.bindings` and `view.model` to set up bindings. This view has its model set during initialization. Why would passing additional context to the template mess up the data binding?

File: ecommerce/static/js/models/course_seats/credit_seat.js [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+            ),
+
+            validation: _.extend({}, CourseSeat.prototype.validation,
+                {
+                    credit_provider: {
+                        required: true,
+                        msg: gettext('All credit seats must have a credit provider.')
+                    },
+                    credit_hours: {
+                        required: true,
```
Comment (clintonb) [inline ecommerce/static/js/models/course_seats/credit_seat.js:25]: You should also validate that the number is greater than 0.

File: ecommerce/static/js/test/specs/models/course_model_spec.js [hunk 24]
```diff
@@ -17,8 +17,8 @@ define([
 
         var model,
             honorSeat = {
-                id: 9,
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/models/course_model_spec.js:20]: Why the ID change?

File: ecommerce/static/js/test/specs/views/course_seat_form_fields/professional_course_seat_form_field_view_spec.js [hunk 25]
```diff
@@ -1,38 +0,0 @@
-define([
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/views/course_seat_form_fields/professional_course_seat_form_field_view_spec.js:1]: Why remove these specs? Having a spec corresponding to the class seems ideal for keeping files organized and relatively short.

File: ecommerce/static/js/utils/utils.js (hunk 19 above)
Comment (clintonb) [inline ecommerce/static/js/utils/utils.js]: `_s.sprintf` will help avoid bugs here.

File: ecommerce/static/js/views/course_form_view.js [hunk 26]
```diff
@@ -287,7 +291,7 @@ define([
                 }
 
                 // Retrieve these after any new renderings.
-                $courseSeats = $courseSeatsContainer.find('.course-seat');
```
Comment (clintonb) [inline ecommerce/static/js/views/course_form_view.js:290]: Any particular reason for this change? We may eventually get rid of Bootstrap, so using the semantic class name is preferable in this case.

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js [hunk 27]
```diff
@@ -35,55 +37,22 @@ define([
             },
 
             className: function () {
-                return 'row course-seat ' + this.seatType;
+                return 'row ' + this.seatType + ' course-seat';
```
Comment (clintonb) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js:40]: Why the change given that the output is the same?

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+                return this;
+            },
+
+            /**
+             * Removes the selected row from the seat table.
+             */
+            removeSeatTableRow: function () {
+                // Remove deleted seat from course product collection.
+                this.course.get('products').remove(this.model);
+                this.undelegateEvents();
```
Comment (clintonb) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js]: `remove()` handles this for you: http://backbonejs.org/#View-remove.

File: ecommerce/static/templates/credit_course_seat_form_field_row.html [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+            min="5" step="1" pattern="\d+" value="<%= price %>">
+    </div>
+</td>
+<td class="credit-hours">
+    <div class="input-group">
+        <div class="input-group-addon">
+            <i class="fa fa-clock-o" aria-hidden="true"></i>
+        </div>
+        <input type="number" class="form-control" id="credit_hours" name="credit_hours"
+            aria-labelledby="credit-hours-label" min="1" max="6" step="1" pattern="\d+" value="">
```
Comment (clintonb) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: Why the max of 6?

Comment (clintonb) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: Also, the `value` attribute can be removed.

File: ecommerce/static/templates/credit_course_seat_form_field_row.html [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+    </div>
+</td>
+<td class="expires">
+    <div class="input-group">
+        <div class="input-group-addon">
+            <i class="fa fa-calendar" aria-hidden="true"></i>
+        </div>
+        <input type="datetime-local" class="form-control" id="expires" name="expires"
+            aria-labelledby="expires-label" value="">
```
Comment (clintonb) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: This, and any other empty value attributes should be removed.

File: ecommerce/static/templates/credit_course_seat_form_field_row.html [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+    </div>
+</td>
+<% if (isRemovable) { %>
+<td>
+    <button class="remove-seat btn btn-danger">
+        <span class="sr">Delete course seat</span>
```
Comment (clintonb) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: Mark for translation.

File: ecommerce/static/templates/credit_course_seat_form_field_row.html (hunk 30 above)
Comment (clintonb) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: There is no "expires-label".

File: ecommerce/static/templates/credit_course_seat_form_field.html [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+        <table class="table table-striped credit-seats">
+            <thead>
+                <tr>
+                    <th id="credit-provider-label"><%= gettext('Credit Provider') %></th>
+                    <th id="price-label"><%= gettext('Price (USD)') %></th>
+                    <th id="credit-hours-label"><%= gettext('Credit Hours') %></th>
+                    <th id="expires-label"><%= gettext('Upgrade Deadline') %></th>
+                    <th>
+                        <button class="add-seat btn btn-primary">
+                            <span class="sr">Add course seat</span>
```
Comment (clintonb) [inline ecommerce/static/templates/credit_course_seat_form_field.html]: Mark for translation.

File: ecommerce/static/js/models/course_seats/credit_seat.js (hunk 23 above)
Comment (rlucioni) [inline ecommerce/static/js/models/course_seats/credit_seat.js:25]: Updated.

File: ecommerce/static/js/test/specs/models/course_model_spec.js (hunk 24 above)
Comment (rlucioni) [inline ecommerce/static/js/test/specs/models/course_model_spec.js:20]: Just a minor reordering so that the IDs proceed more logically in the test data (parent = 7, honor = 8, verified = 9, credit = 10).

File: ecommerce/static/js/test/specs/views/course_seat_form_fields/professional_course_seat_form_field_view_spec.js (hunk 25 above)
Comment (rlucioni) [inline ecommerce/static/js/test/specs/views/course_seat_form_fields/professional_course_seat_form_field_view_spec.js:1]: The specs were testing dead code, which has also been removed.

File: ecommerce/static/js/views/course_form_view.js (hunk 26 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_form_view.js:290]: It's no longer guaranteed that the seat "sections" will all have the `course-seat` class at the top level. Each row in the credit seat table is assigned this class, but the credit seat "section" doesn't have the `course-seat` class. The intention here is to retrieve and hide these seat "sections", not individual course seats.

File: ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js (hunk 27 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/course_seat_form_field_view.js:40]: It's cosmetic, to be consistent with the way the credit seat "section" is assigned its [className](https://github.com/edx/ecommerce/pull/299/files#diff-50d977ab761972b23eb11e40b102324aR29).

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_view.js [hunk 33]
```diff
@@ -0,0 +1,92 @@
...
+
+                // Instantiate new Views handling data binding for each Model in the Collection.
+                this.collection.each( function (seat) {
+                    row = new this.rowView({
+                        model: seat,
+                        isRemovable: false,
+                        course: this.course
+                    });
+
+                    row.render();
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_view.js:63]: It looks like the rowView already calls this on initialize (https://github.com/edx/ecommerce/blob/renzo/credit-create-edit/ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js#L46)

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_view.js [hunk 34]
```diff
@@ -0,0 +1,92 @@
...
+                    $tableBody.append(row.el);
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_view.js]: It would be more performant to store all of the elements in a variable and append them in a single function outside the each loop.

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_view.js [hunk 35]
```diff
@@ -0,0 +1,92 @@
...
+                    seat = new seatClass(),
+                    /*jshint newcap: true */
+                    row = new this.rowView({
+                        model: seat,
+                        isRemovable: true,
+                        course: this.course
+                    }),
+                    $tableBody = this.$el.find('tbody');
+
+                row.render();
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_view.js:87]: It looks like the rowView already calls this on initialize (https://github.com/edx/ecommerce/blob/renzo/credit-create-edit/ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js#L46)

File: ecommerce/static/templates/course_form.html [hunk 36]
```diff
@@ -1,4 +1,4 @@
-<div class="alerts"></div>
+<div class="alerts" tabindex="-1"></div>
```
Comment (AlasdairSwan) [inline ecommerce/static/templates/course_form.html]: There should be an aria-live="polite" attribute here

File: ecommerce/static/templates/credit_course_seat_form_field_row.html [hunk 37]
```diff
@@ -0,0 +1,43 @@
...
+        <i class="fa fa-trash" aria-hidden="true"></i>
+        Delete
```
Comment (AlasdairSwan) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: Mark for translation. Also, with the sr text above this will read as 'Delete course seat Delete". I would recommend moving the sr span above to just after this Delete and updating so line 40 will be

```
Delete<span class="sr"> course seat</span>
```

Translations will make it more interesting than that though :dancers:

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js (hunk 28 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_row_view.js]: Removed.

File: ecommerce/static/templates/credit_course_seat_form_field_row.html (hunk 29 above)
Comment (rlucioni) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: One of the acceptance criteria on the parent story, [XCOM-325](https://openedx.atlassian.net/browse/XCOM-325), is that the user should be able to associate up to 6 credit hours with a credit seat. It seems arbitrary, but I was going by what was written there.

File: ecommerce/static/templates/credit_course_seat_form_field_row.html (hunk 30 above)
Comment (rlucioni) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: Removed.

File: ecommerce/static/templates/credit_course_seat_form_field_row.html (hunk 31 above)
Comment (rlucioni) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: Nice catch. Marked.

File: ecommerce/static/templates/credit_course_seat_form_field.html (hunk 32 above)
Comment (rlucioni) [inline ecommerce/static/templates/credit_course_seat_form_field.html]: Marked.

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_view.js (hunk 33 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_view.js:63]: Whoops, missed that. `rowView` no longer does that.

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_view.js (hunk 34 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_view.js]: Updated.

File: ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_view.js (hunk 35 above)
Comment (rlucioni) [inline ecommerce/static/js/views/course_seat_form_fields/credit_course_seat_form_field_view.js:87]: Fixed above; `rowView` no longer does that.

File: ecommerce/static/templates/course_form.html (hunk 36 above)
Comment (rlucioni) [inline ecommerce/static/templates/course_form.html]: Added.

File: ecommerce/static/templates/credit_course_seat_form_field_row.html (hunk 37 above)
Comment (rlucioni) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: @AlasdairSwan I addressed this by removing the unnecessary "Delete" and marking the screenreader text for translation; I did the same for the seat addition button. The icons are sufficient for sighted users.

File: ecommerce/static/templates/credit_course_seat_form_field_row.html (hunk 29 above)
Comment (rlucioni) [inline ecommerce/static/templates/credit_course_seat_form_field_row.html]: Updated in light of comments on the ticket.

**Comment by rlucioni** [comment]
@clintonb or @AlasdairSwan, any further comments? I've addressed everything up to this point.

**Comment by clintonb** [comment]
:+1:

**Comment by rlucioni** [comment]
@AlasdairSwan if you have further comments, let me know and I'll address them with a follow-up PR.
